File: backend/app/routers/tags.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, distinct
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.video import Tag
from app.services.ingest.tagmap import AUTO_TAG_NAMES

router = APIRouter(prefix="/tags", tags=["tags"])


@router.get("")
async def list_all_tags(
    custom: bool = Query(
        False,
        description=(
            "Drop the automatic category/format tags and return only names the "
            "user typed. The auto ones duplicate the library's kind filter."
        ),
    ),
    db: AsyncSession = Depends(get_db),
) -> list[str]:
    result = await db.execute(select(distinct(Tag.name)).order_by(Tag.name))
    names = [row[0] for row in result.all()]
    if custom:
        names = [n for n in names if n not in AUTO_TAG_NAMES]
    return names

File: frontend/src/lib/api/tags.ts
import { http } from "./client";

export const tagApi = {
  // custom: drop the automatic category/format tags (video, mp4, pdf, …), which
  // only restate the kind filter the library already shows as facets.
  listAll: (custom = false) =>
    http.get<string[]>("/tags", custom ? { params: { custom: true } } : undefined)
      .then(r => r.data),

  add: (videoId: string, name: string) =>
    http.post<string[]>(`/videos/${videoId}/tags`, { name }).then(r => r.data),

  remove: (videoId: string, name: string) =>
    http.delete(`/videos/${videoId}/tags/${name}`).then(r => r.data),
};
